Change detection page for new book (#2188)
bevyengine/bevy-website: content/learn/book/control-flow/change-detection.md

weight = 8
status = 'hidden'
+++

Change detection is an inherent feature of Bevy's Entity Component System (ECS). Each time a
[`Component`] or [`Resource`] is modified, Bevy marks the item as changed. This mechanism helps

> optimize performance by avoiding unnecessary system executions when no relevant changes have

Reviewer: I don't think it avoids system executions, it just filters out items that haven't changed while you're iterating over the query

Author: Changed "system executions" to "calculations".

occurred. Change detection can also be used to trigger actions in response to changes, such as
synchronizing data between two contexts.

In this section, we'll explore how Bevy tracks changes and how you can leverage this
feature in your game development workflow.

## Filtering

> You can configure queries to only iterate over entities where certain components have been modified.

Reviewer: `Changed<T>` and `Added<T>` are not archetype filters. This means that it still needs to iterate over all entities that otherwise match the query, it just only retains items that were changed or added. Excluding possible inlining differences, I believe these should be roughly equivalent:

```rust
fn system1(query: Query<(&MyComponent, &Transform), Changed<Transform>>) {
    for (my_component, transform) in &query {
        todo!()
    }
}

fn system2(query: Query<(&MyComponent, Ref<Transform>)>) {
    for (my_component, transform) in &query {
        if transform.is_changed() {
            todo!()
        }
    }
}
```

IMO this is a pretty important performance detail for users to be aware of. I've seen Bevy's built-in change detection be seriously sub-optimal when you have a small number of changes but a lot of matched entities.

I would like to see this worded slightly differently, and to perhaps have an aside that briefly talks about archetype filters (similar to what Bevy's [docs have](https://dev-docs.bevy.org/bevy/ecs/prelude/struct.Changed.html#time-complexity)). It could also mention `OnAdd` as an (often preferable IME) alternative to `Added<T>`, though we probably haven't covered observers yet in the book, so it can also be left for later.

        }
    }
}

// Using Mut<T> for change detection with mutable access
fn check_and_update_position(mut query: Query<(Entity, Mut<Position>)>) {
    for (entity, mut position) in &mut query {
        if position.is_changed() {
            println!("Entity {:?} position changed to {:?} before update", entity, *position);
            position.x += 1.0; // This will mark the component as changed
        }
    }
}
```

## Resources

The [`Res<T>`] and [`ResMut<T>`] provide the same [`.is_added()`] and [`.is_changed()`] methods as
components do:

```rust
fn detect_changed_score(score: Res<Score>) {
    if score.is_changed() {
        println!("The score changed to {}", score.0);
    }
}
```

## Removed Components

Change detection works differently for removed components, since the component (and possibly the
entity) no longer exists!

To detect when components are removed, you can use the [`RemovedComponents`] event:

```rust
fn detect_removed_position(mut removed: RemovedComponents<Position>) {
    for entity in removed.iter() {
        println!("Entity {:?} just lost its Position component", entity);
    }
}

> ```

Reviewer: IME using `RemovedComponents<T>` is an anti-pattern for the  majority of cases. It can miss component removals when used in `FixedUpdate`, and doesn't let you access the component values. Using `OnRemove` or the `on_remove` hook tends to be preferable, excluding specific cases where you want to quickly process a very large number of removals at once, and don't need the component values.

So I would *at least* link to `OnRemove` here, though again I don't think we have a section on observers yet...

weight = 8
status = 'hidden'
+++

Change detection is built into Bevy's Entity Component System (ECS). Each time a
[`Component`] or [`Resource`] is modified, Bevy marks the item as changed. This mechanism helps
optimize performance by avoiding unnecessary calculations when no relevant changes have
occurred. Change detection can also be used to trigger actions in response to changes, such as
synchronizing data between two contexts.

In this section, we'll explore how Bevy tracks changes and how you can leverage this
feature in your game development workflow.

## Filtering

You can configure queries to filter out entities unless certain components have been modified.

The [`Added<T>`] query filter detects new component instances, either if the component was added to

> an existing entity, or a new entity with that component was spawned.

Reviewer 2: ```suggestion
an existing entity, or a new entity with that component was spawned.
This is also triggered if a component is reinserted on an entity that already had it.
```

by the query.
{% end %}

## Resources

The [`Res<T>`] and [`ResMut<T>`] provide the same [`.is_added()`] and [`.is_changed()`] methods as
components do:

```rust
fn detect_changed_score(score: Res<Score>) {
    if score.is_changed() {
        println!("The score changed to {}", score.0);
    }
}
```

## Removed Components

Change detection works differently for removed components, since the component (and possibly the
entity) no longer exists!

To detect when components are removed, you can use the [`RemovedComponents`] param:

```rust
fn detect_removed_position(mut removed: RemovedComponents<Position>) {
    for entity in removed.iter() {
        println!("Entity {:?} just lost its Position component", entity);
    }
}
```

{% callout(type="warning") %}
It's generally better to use an [`OnRemove`] observer or a component hook to detect removals.
This has a number of advantages over using [`RemovedComponents`]:

- You get access to the component values being removed.
- [`RemovedComponents`] can miss component removals when used in `FixedUpdate`.

{% end %}

Bevy does not provide any API for detecting when resources are removed.

## What gets detected?

> Change detection occurs when a component or resource is mutably dereferenced. [`Mut<T>`] and

Reviewer 2: ```suggestion
Change detection is triggered when a component or resource is mutably dereferenced. [`Mut<T>`] and
```

> [`ResMut<T>`] implement `DerefMut`, which marks the item as changed.

Reviewer 2: ```suggestion
[`ResMut<T>`] implement `DerefMut`, with an implementation that mark the item as changed.
```

only if the new value is different (this requires that the component support `PartialEq`):

```rust
fn update_player_health(mut query: Query<(Entity, &mut Health), With<Player>>) {
    for (entity, mut health) in query.iter_mut() {
        let new_health = (health.0 - 10.0).max(0.0);
        // This marks the component as changed if `new_health` is different.
        health.set_if_neq(Health(new_health));
    }
}
```

Change detection applies to each ECS system separately: a system will see whatever changes occurred
since the last time the system ran. This includes systems that only run sometimes (such as when
using [states] or [run conditions]), so you do not need to worry about "missing" changes.

{% callout(type="info") %}
Internally, Bevy stores an "engine tick count" with every component and resource that marks the
last time that it was updated. When a system calls [`.is_changed()`], it compares the tick count
of the component with the tick count of the last time the system was run.
{% end %}

One thing to beware of is potential 1-frame delay, if the system that is causing the change

> runs after the system that is checking for changes. You may need to pay attention to [run schedules]

Reviewer 2: ```suggestion
runs after the system that is checking for changes. You may need to pay attention to how you [run schedules]
```
